File: Sorting/Python/insertionSort.py
def insertionSort(arr):
    n = len(arr)
    for i in range(n):
        j = i
        while j>0 and arr[j-1]>arr[j]:
            arr[j-1],arr[j]=arr[j],arr[j-1]
            j = j-1

    '''for i in range(n):
            for j in range(i,0,-1):
                if arr[j-1]>arr[j]:
                    arr[j-1],arr[j]=arr[j],arr[j-1]
                else:
                    break'''
    
    return arr

# The worst and average time complexities of Insertion sort is: O(N*N)
# The best case is when all elements are sorted, then the time complexity is: O(N)

def test_insertion_sort():
    # Test Case 1: Random array
    arr1 = [54, 26, 93, 17, 77, 31, 44, 55, 20]
    assert insertionSort(arr1) == sorted(arr1)

    # Test Case 2: Sorted array (ascending)
    arr2 = [10, 20, 30, 40, 50, 60]
    assert insertionSort(arr2) == sorted(arr2)

    # Test Case 3: Sorted array (descending)
    arr3 = [90, 80, 70, 60, 50, 40]
    assert insertionSort(arr3) == sorted(arr3)

    # Test Case 4: Array with duplicate elements
    arr4 = [5, 2, 7, 2, 3, 5, 9, 3]
    assert insertionSort(arr4) == sorted(arr4)

    # Test Case 5: Array with negative numbers
    arr5 = [-10, -5, -3, -8, -15]
    assert insertionSort(arr5) == sorted(arr5)

    # Test Case 6: Empty array
    arr6 = []
    assert insertionSort(arr6) == sorted(arr6)

    # Test Case 7: Array with one element
    arr7 = [42]
    assert insertionSort(arr7) == sorted(arr7)

# Run the test cases
test_insertion_sort()
print("All test cases passed!")

File: Sorting/Java/InsertionSort.java
package Sorting.Java;

public class InsertionSort {
    public static void insertionSort(int[] arr){
        int n = arr.length;
        for(int i=1;i<n;i++){
            /*for(int j=i;j>0&&arr[j]<arr[j-1];j--){
                swap(arr,j,j-1);
            }*/
            int j = i;
            while(j>0 && arr[j]<arr[j-1]){
                swap(arr,j,j-1);
                j--;
            }
        }
    }

    protected static void insertionSortwithoutSwap(int[] arr){
        for(int i=1;i<arr.length;i++){
            int key = arr[i];
            int j = i - 1;

            while(j>=0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = key;
        }
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void insertionSortRecursive(int[] arr, int n) {
        if (n <= 1) {
            return;
        }
        insertionSortRecursive(arr, n - 1);
        int last = arr[n - 1];
        int j = n - 2;
        while (j >= 0 && arr[j] > last) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = last;
    }

    public static void printArray(int[] nums) {
        for(int i: nums) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = {5, 2, 9, 1, 5, 6};
        insertionSort(arr);
        System.out.print("Sorted array: ");
        printArray(arr);
        int[] arrRecursive = {5, 2, 9, 1, 5, 6};
        insertionSortRecursive(arrRecursive, arrRecursive.length);
        System.out.print("Sorted array using recursion: ");
        printArray(arrRecursive);
        int[] arrNoSwap = {8,3,7,4,6,2,5,9};
        insertionSortwithoutSwap(arrNoSwap);
        System.out.print("Sorted array without swapping: ");
        printArray(arrNoSwap);
    }
}


// Time complexity for Insertion sort is: O(n^2) in the worst and average cases (O(n^2/4))
// Best case time complexity is: O(n) as no swapping is needed
// Space Complexity is: O(1)